api/rql/interfaces.go

```diff
@@ -0,0 +1,93 @@
+package rql
+
+import (
+	"time"
+
+	"github.com/puppetlabs/wash/plugin"
+	"github.com/shopspring/decimal"
+)
+
+// Primary is the interface implemented by all the primaries. A primary's
+// domain is the set of all entries that it applies to. The domain can be
+// specified at the instance-level (EntryInDomain), the schema-level
+// (EntrySchemaInDomain) or both.
+//
+// Note that a primary can either be an EntryPredicate, an EntrySchemaPredicate,
+// both or neither. In practice, the RQL will use EvalEntrySchema when pruning the
+// stree and EvalEntry when filtering the entries. EntryInDomain and EntrySchemaInDomain
+// are only needed to correctly negate the primaries (for without it, strict negation
+// would return true for entries that are outside the primary's domain).
+//
+// For a given primary p, here are the possible scenarios for EvalEntrySchema
+// and EvalEntry (including negation). Note that p.EntrySchemaInDomain and
+// p.EvalEntrySchema can assume that s != nil.
+//   * If p implements EntrySchemaPredicate, then for RQL
+//         EvalEntrySchema(s) == p.EntrySchemaInDomain(s) && p.EvalEntrySchema(s)
```

Comment:
I'm not clear what the evaluation order is in this expression.

Reply:
Should be `InDomain` first then the `Eval` method w/ appropriate short-circuiting. I'll clarify that.

Reply:
Updated.

Reply:
Parentheses would have also worked.